fix uws polling

e3dio opened the pull request:
### The kind of change this PR does introduce

* [x] a bug fix
* [ ] a new feature
* [ ] an update to the documentation
* [x] a code change that improves performance
* [ ] other

### Current behaviour
Polling error, see https://github.com/socketio/socket.io/discussions/4281

### New behaviour
Prevent error, also improves performance: reduces copies from 2 to 0 or 1, replaces Object.keys(headers).forEach, removes nested inner functions

e3dio commented:
I don't use Socket.io and am not familiar with building and testing it so this is not tested at all, should work but should be reviewed and tested

darrachequesne commented on lib/transports-uws/polling.ts:
```diff
@@ -122,6 +122,18 @@ export class Polling extends Transport {
       return;
     }
 
+    const contentLengthHeader = Number(req.headers["content-length"]);
```
If the content is chunked (`Transfer-Encoding: chunked`), won't the `Content-Length` header be undefined there?

e3dio replied:
Chunked POST's from the client are not supported by uWS, see https://github.com/uNetworking/uWebSockets.js/issues/669 the Socket.io client needs to not use it for this case

darrachequesne replied:
OK, thanks for the link. But in that case, what happens in the issue? Does it throw on the first chunk?

e3dio replied:
I see there is specific http code 411 for Length Required, I will update that https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/411

e3dio replied:
Ok I tested this with new Node v17.5 `fetch` using `{ body: createReadStream() }` to send `transfer-encoding: chunked` request, uWS.js successfully responds with status `411 Length Required` when request is missing `content-length` header

e3dio commented:
The `onDataRequest` function is creating many inner functions including `cleanup` `onClose` `onEnd`, these get created every request and need to be garbage collected every request, it is best to remove or move those outside of hot path. Also it is doing `Object.keys(headers).forEach` when it should be doing `for(let key in headers)` which avoids creating new object. I did not update these to simplify showing how to fix current bug. You need to be thinking about performance when coding servers

e3dio commented:
I made update to this, I changed existing  `Object.keys(headers).forEach` to `for (let key in headers)` and I removed the nested inner function `cleanup`, from my test on v8 declaring inner functions in a hot path does not hurt unless you nest them together like `cleanup` and `onClose` was doing, so should be good now

e3dio commented:
I saw this comment from @darrachequesne on recent commit:

> On Safari, a POST request would result in two chunks (the first being empty), and threw the following error:
> > TypeError: Cannot perform %TypedArray%.prototype.set on a detached ArrayBuffer

> Which is a bit weird, because it seems µWebSockets.js does not support chunked content: https://github.com/uNetworking/uWebSockets.js/issues/669

A normal POST does not use `transfer-encoding: chunked`, chunked is for when you don't know size of content you are sending, like streaming dynamic content, and is not related to this issue. This PR fixes any POST above a certain size that requires more than 1 onData event, as data gets split into chunks to send across network. `transfer-encoding: chunked` is a different protocol for sending data not related to this issue, you are mixing things up, you don't need to be accepting that type of stream for this endpoint

e3dio commented:
You are lacking basic understanding of basic HTTP concepts

darrachequesne commented:
Merged as https://github.com/socketio/engine.io/commit/33674403084c329dc6ad026c4122333a6f8a9992. Thanks.
